### PR TITLE
[Feature] partition prune support of list partition

### PR DESCRIPTION
**Pull request by hiliuxg**
issue #4035

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Problem Summary(Required) ：
olap table partition prune support of list partition

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 42 / 45 (93.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java|42|45|93.33%|[157, 158, 160]|

**Review by wyb — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java [hunk 1]
```diff
@@ -84,8 +97,70 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+
+        List<ScalarOperator> scalarOperatorList = Utils.extractConjuncts(olapScanOperator.getPredicate());
+        PartitionPruner partitionPruner = new ListPartitionPruner(columnToPartitionValuesMap,
+                columnToNullPartitions, scalarOperatorList);
+        try {
+            return partitionPruner.prune();
+        } catch (AnalysisException e) {
+            LOG.warn("PartitionPrune Failed. ", e);
+        }
+        return null;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:160]: better way is return all partition?

Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:160]: better way is return all partition?

Comment (hiliuxg) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PartitionPruneRule.java:160]: the rangePartitionPrune has return null partitions when throw exception. better to keep the same with the rang partition prune ？

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**